{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56aff7fb",
   "metadata": {},
   "source": [
    "# Data formats\n",
    "\n",
    "Chemical compounds are usually stored in databases using the SMILES format. Usually CSV files are used for this purpose, where we have a `smiles` column, and other columns are molecular descriptors or experimental data. Sometimes we only get SMILES, in which case such files have the `.smi` extension, and molecules are saved as one SMILES string per line.\n",
    "\n",
    "## Conformations and force fields\n",
    "\n",
    "SMILES does not describe the spatial arrangement of atoms in the 3D space, which is called a **molecular conformation**. Conformations are produced by rotating single bonds in order to find an energetically stable pose. Multiple conformations are possible, as many conformations can be stable. We can employ computer tools to find conformations close to the real ones. We can use **force fields**, which are optimization methods that iteratively modify the conformation of a compound to minimize an energy function.\n",
    "\n",
    "## 3D data formats for chemical compounds\n",
    "\n",
    "If we need to store a compounds along with its conformation, the best widely used formats are MOL2 or SDF. These data formats enable storing information about atom positions, which is shown in the example below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bc3dd03d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import py3Dmol\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import AllChem\n",
    "\n",
    "from ipywidgets import interact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "23775077",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>smiles</th>\n",
       "      <th>zinc_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CC(=O)Nc1ccc(S(=O)(=O)N(C)CCO)cc1</td>\n",
       "      <td>29138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCOc1ccc(C(=O)N[C@@H](C(=O)O)[C@H](C)O)cc1</td>\n",
       "      <td>344024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CNC(=O)N[C@@H]1N(C)C(=O)N[C@H](C(C)C)C1(C)C</td>\n",
       "      <td>517650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cc1oc2nc(SCC(=O)NCCO)nc(N)c2c1C</td>\n",
       "      <td>2259245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O=C(O)CCCNC1=NS(=O)(=O)c2ccccc21</td>\n",
       "      <td>2637193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114001</th>\n",
       "      <td>O=C(c1nc[nH]c1C(F)(F)F)N1CCc2nn[nH]c2C1</td>\n",
       "      <td>828942673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114002</th>\n",
       "      <td>CNC(=O)CNC(=O)[C@H]1Cc2cc(Cl)ccc2O1</td>\n",
       "      <td>25266002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114003</th>\n",
       "      <td>CCOC1CC(CC(=O)N(C)[C@H]2C[C@@H](C(N)=O)C2)C1</td>\n",
       "      <td>863358233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114004</th>\n",
       "      <td>CNc1ccccc1C(=O)NC[C@@]1(O)CCO[C@@H]1C</td>\n",
       "      <td>227184162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114005</th>\n",
       "      <td>Cc1noc(C)c1S(=O)(=O)NCCCN(C)C</td>\n",
       "      <td>7374622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114006 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              smiles    zinc_id\n",
       "0                  CC(=O)Nc1ccc(S(=O)(=O)N(C)CCO)cc1      29138\n",
       "1         CCOc1ccc(C(=O)N[C@@H](C(=O)O)[C@H](C)O)cc1     344024\n",
       "2        CNC(=O)N[C@@H]1N(C)C(=O)N[C@H](C(C)C)C1(C)C     517650\n",
       "3                    Cc1oc2nc(SCC(=O)NCCO)nc(N)c2c1C    2259245\n",
       "4                   O=C(O)CCCNC1=NS(=O)(=O)c2ccccc21    2637193\n",
       "...                                              ...        ...\n",
       "114001       O=C(c1nc[nH]c1C(F)(F)F)N1CCc2nn[nH]c2C1  828942673\n",
       "114002           CNC(=O)CNC(=O)[C@H]1Cc2cc(Cl)ccc2O1   25266002\n",
       "114003  CCOC1CC(CC(=O)N(C)[C@H]2C[C@@H](C(N)=O)C2)C1  863358233\n",
       "114004         CNc1ccccc1C(=O)NC[C@@]1(O)CCO[C@@H]1C  227184162\n",
       "114005                 Cc1noc(C)c1S(=O)(=O)NCCCN(C)C    7374622\n",
       "\n",
       "[114006 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zinc_df = pd.read_csv('sample.smi', sep=' ')\n",
    "zinc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "55762dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def optimize_conformation(mol):\n",
    "    mol = Chem.AddHs(mol)  # Adds hydrogens to make optimization more accurate\n",
    "    AllChem.EmbedMolecule(mol)  # Adds 3D positions\n",
    "    AllChem.MMFFOptimizeMolecule(mol)  # Improves the 3D positions using a force-field method\n",
    "    return mol\n",
    "\n",
    "\n",
    "@interact\n",
    "def show_molecule(compound_idx=(0, len(zinc_df), 1)):\n",
    "    smiles = zinc_df.iloc[compound_idx].smiles\n",
    "    if not smiles:\n",
    "        return\n",
    "    mol = Chem.MolFromSmiles(smiles)\n",
    "    if not mol:\n",
    "        return\n",
    "    mol = optimize_conformation(mol)\n",
    "    mblock = Chem.MolToMolBlock(mol, kekulize=False)  # Produces a MOL block, which can be stored in a .mol file\n",
    "\n",
    "    view = py3Dmol.view(width=800, height=500)\n",
    "    view.addModel(Chem.MolToMolBlock(mol, kekulize=True), 'mol')\n",
    "    view.setStyle({'model': -1}, {\"stick\": {'color': 'white', 'radius': 0.15}, \"sphere\": {'radius': .4}})\n",
    "    view.zoomTo({'model': -1})\n",
    "    view.show()\n",
    "    \n",
    "    print('-' * 100)\n",
    "    print(mblock)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a738d274",
   "metadata": {},
   "source": [
    "# Proteins\n",
    "\n",
    "Proteins are macromolecules (meaning they are big) that have various functions in the organism. Some proteins are involved in metabolism (mostly enzymes), other ones are involved in tranducing signals (G-protein-coupled receptors), and finally some take part in transporting ions (ion channels). Thus, proteins are the most popular target for small molecule drugs.\n",
    "\n",
    "Proteins are **polymers**, which means they consist of smaller fragments called **amino acids**:\n",
    "\n",
    "![](https://upload.wikimedia.org/wikipedia/commons/thumb/0/0f/Amino_acids.png/924px-Amino_acids.png)\n",
    "\n",
    "![](https://upload.wikimedia.org/wikipedia/commons/3/32/SIMPLIFIED_PEPTIDE_CHAIN.png)\n",
    "\n",
    "Amino acids have different properties and influence the shape (conformation) of the molecule. Because proteins have a linear structure and consist of repeated amino acid blocks, we can store them as a sequence of letters corresponding to the consecutive amino acids, e.g. using the FASTA format. This is also called a **primary protein structure**.\n",
    "\n",
    "```\n",
    ">6NL3_1|Chain A|Cytochrome c oxidase assembly factor 6 homolog|Homo sapiens (9606)\n",
    "GSMAAPSMKERQVCWGARDEYWKCLDENLEDASQCKKLRSSFESSCPQQWIKYFDKRRDYLKFKEKFEAGQFEPSETTAKS\n",
    "```\n",
    "\n",
    "## Protein folding\n",
    "\n",
    "\n",
    "Protein conformations are a more complex problem than conformations of small molecules. The number of atoms and rotatable bonds in proteins is massive, so that the force-field calculation programs work poorly for proteins. Proteins in the organism are made sequentially and instantly undergo folding under well-defined physiological conditions, often with the help of other proteins, so-called chaperones. Therefore, the folding effect is very precise and deterministic despite the many points of freedom.\n",
    "\n",
    "For a long time, the problem of protein folding was unattainable for computer methods, and until now the traditional method of discovering protein conformations is **crystallization**. The crystallized proteins can then be \"photographed\" to obtain a 3D model with high accuracy. Nowadays, computer methods are already achieving accuracy that until recently was thought to be impossible. DeepMind has developed the **AlphaFold** algorithm based on deep learning, which is able to predict 3D structures with an accuracy of 0.96 Å r.m.s.d.95 ( The next solution in the **CASP14** competition had a score of 2.8 Å r.m.s.d.95). DeepMind has started releasing prediction results for proteins whose crystals were previously unknown ([AlphaFold Protein Structure Database](https://alphafold.ebi.ac.uk/)). The AlphaFold implementation has been made open source, and predictions for new protein sequences can be made in-house using the provided [Colab notebook](https://colab.sandbox.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb).\n",
    "\n",
    "![](https://lh3.googleusercontent.com/J5RC0j1DeUWaNc5h6nGwJwsQSLUuTXINP6we2ymLJ_WUg9bH-hvfvI8WVFeghN-_YR69MryNK5O2rFcVNwz9PZePpBtLdwdshCGzLdM=w2048-rw-v1)\n",
    "\n",
    "## The PDB database\n",
    "\n",
    "[PDB](https://www.rcsb.org/) is a database of protein structures, where we can find crystals of the biological targets that we are interested in. The most popular format of protein structures is in fact `.pdb`, a fragment of which is presented below ([CYP 3A4](https://www.rcsb.org/structure/6MA7))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "63486b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "@interact\n",
    "def show_protein(style=['balls & sticks', 'cartoon', 'surface']):\n",
    "    view = py3Dmol.view(width=800, height=500, query='pdb:6MA7')\n",
    "    if style == 'cartoon':\n",
    "        view.setStyle({'model': -1}, {\"cartoon\": {'color': 'cyan'}})\n",
    "    elif style == 'balls & sticks':\n",
    "        view.setStyle({'model': -1}, {\"stick\": {'color': 'white', 'radius': 0.15}, \"sphere\": {'radius': .4}})\n",
    "    elif style == 'surface':\n",
    "        view.setStyle({'model': -1}, {})\n",
    "        view.addSurface('VDW')\n",
    "\n",
    "    view.zoomTo({'model': -1})\n",
    "    view.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd7df37",
   "metadata": {},
   "source": [
    "```\n",
    "...\n",
    "ATOM      1  N   GLY A  26     -34.317   8.706 -14.755  1.00128.59           N  \n",
    "ANISOU    1  N   GLY A  26    16562  13800  18498   2794    437   1597       N  \n",
    "ATOM      2  CA  GLY A  26     -33.904   7.326 -14.563  1.00126.57           C  \n",
    "ANISOU    2  CA  GLY A  26    16140  13908  18043   2542    375   1417       C  \n",
    "ATOM      3  C   GLY A  26     -32.608   6.995 -15.277  1.00128.93           C  \n",
    "ANISOU    3  C   GLY A  26    16532  14226  18228   2319    463   1607       C  \n",
    "ATOM      4  O   GLY A  26     -32.168   5.852 -15.299  1.00117.11           O  \n",
    "ANISOU    4  O   GLY A  26    14921  13024  16553   2131    417   1500       O  \n",
    "ATOM      5  N   THR A  27     -31.996   8.010 -15.878  1.00144.25           N  \n",
    "ANISOU    5  N   THR A  27    18681  15832  20295   2349    618   1906       N  \n",
    "ATOM      6  CA  THR A  27     -30.716   7.829 -16.548  1.00149.54           C  \n",
    "ANISOU    6  CA  THR A  27    19431  16468  20918   2141    773   2119       C  \n",
    "ATOM      7  C   THR A  27     -29.647   8.652 -15.842  1.00161.94           C  \n",
    "ANISOU    7  C   THR A  27    21063  17579  22889   1854    953   2076       C  \n",
    "ATOM      8  O   THR A  27     -29.018   9.523 -16.456  1.00167.94           O  \n",
    "ANISOU    8  O   THR A  27    21978  18006  23825   1843   1158   2384       O  \n",
    "ATOM      9  CB  THR A  27     -30.818   8.232 -18.017  1.00142.27           C  \n",
    "ANISOU    9  CB  THR A  27    18706  15556  19794   2424    845   2558       C  \n",
    "ATOM     10  OG1 THR A  27     -31.069   9.640 -18.105  1.00145.88           O  \n",
    "ANISOU   10  OG1 THR A  27    19354  15572  20501   2606    963   2775       O  \n",
    "ATOM     11  CG2 THR A  27     -31.950   7.475 -18.695  1.00132.20           C  \n",
    "ANISOU   11  CG2 THR A  27    17362  14727  18142   2738    591   2545       C  \n",
    "...\n",
    "HETATM 3735  CHA HEM A 601     -15.300 -19.651  -9.279  1.00 42.50           C  \n",
    "ANISOU 3735  CHA HEM A 601     4140   6322   5688     50   -273   -757       C  \n",
    "HETATM 3736  CHB HEM A 601     -16.180 -22.481 -13.088  1.00 45.45           C  \n",
    "ANISOU 3736  CHB HEM A 601     4576   6920   5771    211     98   -793       C  \n",
    "HETATM 3737  CHC HEM A 601     -16.854 -26.224 -10.126  1.00 35.75           C  \n",
    "ANISOU 3737  CHC HEM A 601     3757   5260   4565    363    -24   -922       C  \n",
    "HETATM 3738  CHD HEM A 601     -15.649 -23.507  -6.308  1.00 43.62           C  \n",
    "ANISOU 3738  CHD HEM A 601     4865   6249   5459    446   -422   -814       C  \n",
    "HETATM 3739  C1A HEM A 601     -15.509 -20.056 -10.578  1.00 42.15           C  \n",
    "ANISOU 3739  C1A HEM A 601     4058   6356   5602     54   -122   -714       C  \n",
    "...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c531c0f5",
   "metadata": {},
   "source": [
    "# Molecular docking\n",
    "\n",
    "Molecular docking is a tool that facilitates the examination of interactions between a small molecule and its protein target. The surface of a protein has visible cavities (**binding pockets**) into which a molecule (also called a **ligand**) can enter to modulate the function of a given protein. Molecular docking tools aim to place the ligand in the most energetically favorable position in the protein's binding pocket. An energy function is defined, which is optimized by repositioning the ligand in the pocket and rotating the rotatable bonds. The conformation obtained by this process is called the **active conformation**.\n",
    "\n",
    "Docking is also used in **virtual screening** to find molecules that bind well to a protein. This simulation suffers from considerable amount of error, but is very effective in filtering out molecules that bind poorly. A good docking score does not yet indicate the activity of the compound, but is an approximation of it."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83bba996",
   "metadata": {},
   "source": [
    "**Exercise:** Dock molecules to the BACE crystal. Do docking scores correlate with compound activity?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "fae03a6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mol</th>\n",
       "      <th>CID</th>\n",
       "      <th>Class</th>\n",
       "      <th>Model</th>\n",
       "      <th>pIC50</th>\n",
       "      <th>MW</th>\n",
       "      <th>AlogP</th>\n",
       "      <th>HBA</th>\n",
       "      <th>HBD</th>\n",
       "      <th>RB</th>\n",
       "      <th>...</th>\n",
       "      <th>PEOE6 (PEOE6)</th>\n",
       "      <th>PEOE7 (PEOE7)</th>\n",
       "      <th>PEOE8 (PEOE8)</th>\n",
       "      <th>PEOE9 (PEOE9)</th>\n",
       "      <th>PEOE10 (PEOE10)</th>\n",
       "      <th>PEOE11 (PEOE11)</th>\n",
       "      <th>PEOE12 (PEOE12)</th>\n",
       "      <th>PEOE13 (PEOE13)</th>\n",
       "      <th>PEOE14 (PEOE14)</th>\n",
       "      <th>canvasUID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O1CC[C@@H](NC(=O)[C@@H](Cc2cc3cc(ccc3nc2N)-c2c...</td>\n",
       "      <td>BACE_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Train</td>\n",
       "      <td>9.154901</td>\n",
       "      <td>431.56979</td>\n",
       "      <td>4.4014</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>53.205711</td>\n",
       "      <td>78.640335</td>\n",
       "      <td>226.85541</td>\n",
       "      <td>107.43491</td>\n",
       "      <td>37.133846</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.980170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fc1cc(cc(F)c1)C[C@H](NC(=O)[C@@H](N1CC[C@](NC(...</td>\n",
       "      <td>BACE_2</td>\n",
       "      <td>1</td>\n",
       "      <td>Train</td>\n",
       "      <td>8.853872</td>\n",
       "      <td>657.81073</td>\n",
       "      <td>2.6412</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>73.817162</td>\n",
       "      <td>47.171600</td>\n",
       "      <td>365.67694</td>\n",
       "      <td>174.07675</td>\n",
       "      <td>34.923889</td>\n",
       "      <td>7.980170</td>\n",
       "      <td>24.148668</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.663788</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S1(=O)(=O)N(c2cc(cc3c2n(cc3CC)CC1)C(=O)N[C@H](...</td>\n",
       "      <td>BACE_3</td>\n",
       "      <td>1</td>\n",
       "      <td>Train</td>\n",
       "      <td>8.698970</td>\n",
       "      <td>591.74091</td>\n",
       "      <td>2.5499</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>70.365707</td>\n",
       "      <td>47.941147</td>\n",
       "      <td>192.40652</td>\n",
       "      <td>255.75255</td>\n",
       "      <td>23.654478</td>\n",
       "      <td>0.230159</td>\n",
       "      <td>15.879790</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.663788</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S1(=O)(=O)C[C@@H](Cc2cc(O[C@H](COCC)C(F)(F)F)c...</td>\n",
       "      <td>BACE_4</td>\n",
       "      <td>1</td>\n",
       "      <td>Train</td>\n",
       "      <td>8.698970</td>\n",
       "      <td>591.67828</td>\n",
       "      <td>3.1680</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>56.657166</td>\n",
       "      <td>37.954151</td>\n",
       "      <td>194.35304</td>\n",
       "      <td>202.76335</td>\n",
       "      <td>36.498634</td>\n",
       "      <td>0.980913</td>\n",
       "      <td>8.188327</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.385181</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S1(=O)(=O)N(c2cc(cc3c2n(cc3CC)CC1)C(=O)N[C@H](...</td>\n",
       "      <td>BACE_5</td>\n",
       "      <td>1</td>\n",
       "      <td>Train</td>\n",
       "      <td>8.698970</td>\n",
       "      <td>629.71283</td>\n",
       "      <td>3.5086</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>78.945702</td>\n",
       "      <td>39.361153</td>\n",
       "      <td>179.71288</td>\n",
       "      <td>220.46130</td>\n",
       "      <td>23.654478</td>\n",
       "      <td>0.230159</td>\n",
       "      <td>15.879790</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.100143</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 595 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 mol     CID  Class  Model  \\\n",
       "0  O1CC[C@@H](NC(=O)[C@@H](Cc2cc3cc(ccc3nc2N)-c2c...  BACE_1      1  Train   \n",
       "1  Fc1cc(cc(F)c1)C[C@H](NC(=O)[C@@H](N1CC[C@](NC(...  BACE_2      1  Train   \n",
       "2  S1(=O)(=O)N(c2cc(cc3c2n(cc3CC)CC1)C(=O)N[C@H](...  BACE_3      1  Train   \n",
       "3  S1(=O)(=O)C[C@@H](Cc2cc(O[C@H](COCC)C(F)(F)F)c...  BACE_4      1  Train   \n",
       "4  S1(=O)(=O)N(c2cc(cc3c2n(cc3CC)CC1)C(=O)N[C@H](...  BACE_5      1  Train   \n",
       "\n",
       "      pIC50         MW   AlogP  HBA  HBD  RB  ...  PEOE6 (PEOE6)  \\\n",
       "0  9.154901  431.56979  4.4014    3    2   5  ...      53.205711   \n",
       "1  8.853872  657.81073  2.6412    5    4  16  ...      73.817162   \n",
       "2  8.698970  591.74091  2.5499    4    3  11  ...      70.365707   \n",
       "3  8.698970  591.67828  3.1680    4    3  12  ...      56.657166   \n",
       "4  8.698970  629.71283  3.5086    3    3  11  ...      78.945702   \n",
       "\n",
       "   PEOE7 (PEOE7)  PEOE8 (PEOE8)  PEOE9 (PEOE9)  PEOE10 (PEOE10)  \\\n",
       "0      78.640335      226.85541      107.43491        37.133846   \n",
       "1      47.171600      365.67694      174.07675        34.923889   \n",
       "2      47.941147      192.40652      255.75255        23.654478   \n",
       "3      37.954151      194.35304      202.76335        36.498634   \n",
       "4      39.361153      179.71288      220.46130        23.654478   \n",
       "\n",
       "   PEOE11 (PEOE11)  PEOE12 (PEOE12)  PEOE13 (PEOE13)  PEOE14 (PEOE14)  \\\n",
       "0         0.000000         7.980170              0.0         0.000000   \n",
       "1         7.980170        24.148668              0.0        24.663788   \n",
       "2         0.230159        15.879790              0.0        24.663788   \n",
       "3         0.980913         8.188327              0.0        26.385181   \n",
       "4         0.230159        15.879790              0.0        26.100143   \n",
       "\n",
       "   canvasUID  \n",
       "0          1  \n",
       "1          2  \n",
       "2          3  \n",
       "3          4  \n",
       "4          5  \n",
       "\n",
       "[5 rows x 595 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bace_path = 'https://deepchemdata.s3-us-west-1.amazonaws.com/datasets/bace.csv'\n",
    "df = pd.read_csv(bace_path)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a0351a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "docking_scores = []\n",
    "for i, row in df.iterrows():\n",
    "    smiles = row.mol\n",
    "    mol = Chem.MolFromSmiles(smiles)\n",
    "    docking_score = ...  # compute docking score, watch out for non-docking compounds\n",
    "    docking_scores.append(docking_score)\n",
    "df['docking_score'] = docking_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "2496a4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sns.scatterplot(data=df, x='pIC50', y='docking_score')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55f60565",
   "metadata": {},
   "source": [
    "### Molecular docking code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "356a760f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dock_molecule(mol, program='smina'):\n",
    "    Chem.MolToMolFile(mol, f'molecule.mol')\n",
    "    !obabel -imol molecule.mol -omol2 -O molecule.mol2\n",
    "    os.remove(f'molecule.mol')\n",
    "    \n",
    "    # Dock the molecule stored in file molecule.mol2\n",
    "    !{program} -r 4ivt.pdb -l molecule.mol2 --center_x 22.312222 --center_y 23.862839 --center_z 0.34194124 --size_x 30 --size_y 30 --size_z 30 --exhaustiveness 8 --out molecule_docked.mol2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3c27058a",
   "metadata": {},
   "outputs": [],
   "source": [
    "smiles = df.mol.iloc[0]\n",
    "mol = Chem.MolFromSmiles(smiles)\n",
    "optimize_conformation(mol)\n",
    "dock_molecule(mol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a5904292",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Score only to see the scoring function components\n",
    "import re\n",
    "\n",
    "output = !smina -r 4ivt.pdb -l molecule_docked.mol2 --score_only\n",
    "re.findall(r'Affinity:\\s*(\\-?[\\d\\.]+)', '\\n'.join(output))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "9ea96ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"4ivt.pdb\") as ifile:\n",
    "    system = \"\".join([x for x in ifile])\n",
    "\n",
    "if os.path.exists(f\"molecule_docked.mol2\"):\n",
    "    with open(f\"molecule_docked.mol2\") as ifile:\n",
    "        mol = \"\".join([x for x in ifile])\n",
    "        mols = ['@<TRIPOS>MOLECULE' + conf for conf in mol.split('@<TRIPOS>MOLECULE') if conf]\n",
    "else:\n",
    "    mols = None\n",
    "\n",
    "view = py3Dmol.view(width=800, height=500)\n",
    "view.addModelsAsFrames(system)\n",
    "view.setStyle({'model': -1}, {\"cartoon\": {'color': 'cyan'}})\n",
    "if mols:\n",
    "    view.addModel(mols[0], 'mol')\n",
    "    view.setStyle({'model': -1}, {\"stick\": {'color': 'white', 'radius': 0.15}, \"sphere\": {'radius': .4}})\n",
    "# view.addBox({'center': {'x': pocket_center[0], 'y': pocket_center[1], 'z': pocket_center[2]}, 'dimensions': {'w': 30, 'h': 30, 'd': 30}, 'wireframe': True})\n",
    "view.zoomTo({'model': -1})\n",
    "view.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d19b5235",
   "metadata": {},
   "source": [
    "## Detailed analysis of docked compounds\n",
    "\n",
    "If we have our compounds docked, we can inspect interactions between the ligand and the protein. To that end, we can use the [ProLIF](https://prolif.readthedocs.io/en/latest/notebooks/visualisation.html) package. For example, we can use **interaction fingerprints** to train machine learning models. These fingerprints describe which ligand atoms interact with which amino acids, and what the types of interactions are.\n",
    "\n",
    "Another important example of how we can use 3D shape of the docked compounds is to compute a **pharmacophore**, which is a set of 3D features that describe key interactions with the target. If we can find a set of interactions that are important for the ligand binding, we can propose a pharmacophoric hypothesis, which can be used to filter big databases of molecules in order to find so called **hits**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4cc0c4ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem import ChemicalFeatures\n",
    "from rdkit import RDConfig\n",
    "import numpy as np\n",
    "\n",
    "fdefName = os.path.join(RDConfig.RDDataDir,'BaseFeatures.fdef')\n",
    "factory = ChemicalFeatures.BuildFeatureFactory(fdefName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2275da87",
   "metadata": {},
   "outputs": [],
   "source": [
    "mol = Chem.MolFromMol2File('molecule_docked.mol2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4cb95421",
   "metadata": {},
   "outputs": [],
   "source": [
    "mol = Chem.AddHs(mol, addCoords=True)\n",
    "\n",
    "view = py3Dmol.view(width=800, height=500)\n",
    "view.addModelsAsFrames(system)\n",
    "view.setStyle({'model': -1}, {\"cartoon\": {'color': 'cyan'}})\n",
    "\n",
    "view.addModel(Chem.MolToMolBlock(mol, kekulize=True), 'mol')\n",
    "view.setStyle({'model': -1}, {\"stick\": {'color': 'white', 'radius': 0.15}, \"sphere\": {'radius': .4}})\n",
    "view.zoomTo({'model': -1})\n",
    "\n",
    "feats = factory.GetFeaturesForMol(mol)\n",
    "for feat in feats:\n",
    "    atom_ids = feat.GetAtomIds()\n",
    "    phtype = feat.GetType()\n",
    "    if phtype == 'SingleAtomDonor':\n",
    "        phcolor = 'blue'\n",
    "    elif phtype == 'SingleAtomAcceptor':\n",
    "        phcolor = 'red'\n",
    "    elif phtype.startswith('Arom'):\n",
    "        phcolor = 'green'\n",
    "    else:\n",
    "        continue\n",
    "        print(phtype)\n",
    "        phcolor = 'grey'\n",
    "    pos = mol.GetConformer().GetPositions()[np.array(atom_ids)].mean(axis=0)\n",
    "    view.addSphere({'center': {'x': pos[0], 'y': pos[1], 'z': pos[2]}, \n",
    "                    'radius': 1.5, 'color': phcolor, 'alpha': 1.0, 'wireframe': True});\n",
    "\n",
    "view.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "010b7294",
   "metadata": {},
   "outputs": [],
   "source": [
    "import prolif as plf\n",
    "from rdkit import Chem\n",
    "\n",
    "mol = Chem.MolFromPDBFile(\"4ivt.pdb\", removeHs=False)\n",
    "pmol = plf.Molecule(mol)\n",
    "\n",
    "with open('molecule_docked.mol2', 'r') as file:\n",
    "    mol_blocks = ['@<TRIPOS>MOLECULE'+block for block in file.read().split('@<TRIPOS>MOLECULE')][1:]\n",
    "    mol = Chem.MolFromMol2Block(mol_blocks[0])\n",
    "lmol = plf.Molecule.from_rdkit(mol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f70d7d1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = plf.Fingerprint()\n",
    "fp.run_from_iterable([lmol], pmol)\n",
    "df = fp.to_dataframe(return_atoms=True)\n",
    "df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f3106289",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit import Chem\n",
    "from rdkit import Geometry\n",
    "\n",
    "\n",
    "def get_ring_centroid(mol, index):\n",
    "    # find ring using the atom index\n",
    "    Chem.SanitizeMol(mol, Chem.SanitizeFlags.SANITIZE_SETAROMATICITY)\n",
    "    ri = mol.GetRingInfo()\n",
    "    for r in ri.AtomRings():\n",
    "        if index in r:\n",
    "            break\n",
    "    else:\n",
    "        raise ValueError(\n",
    "            \"No ring containing this atom index was found in the given molecule\"\n",
    "        )\n",
    "    # get centroid\n",
    "    coords = mol.xyz[list(r)]\n",
    "    ctd = plf.utils.get_centroid(coords)\n",
    "    return Geometry.Point3D(*ctd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b6249d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "import py3Dmol\n",
    "\n",
    "colors = {\n",
    "    \"Hydrophobic\": \"green\",\n",
    "    \"HBAcceptor\": \"cyan\",\n",
    "    \"HBDonor\": \"cyan\",\n",
    "    \"XBDonor\": \"orange\",\n",
    "    \"XBAcceptor\": \"orange\",\n",
    "    \"Cationic\": \"red\",\n",
    "    \"Anionic\": \"blue\",\n",
    "    \"PiStacking\": \"purple\",\n",
    "}\n",
    "\n",
    "# JavaScript functions\n",
    "resid_hover = \"\"\"function(atom,viewer) {{\n",
    "    if(!atom.label) {{\n",
    "        atom.label = viewer.addLabel('{0}:'+atom.atom+atom.serial,\n",
    "            {{position: atom, backgroundColor: 'mintcream', fontColor:'black'}});\n",
    "    }}\n",
    "}}\"\"\"\n",
    "hover_func = \"\"\"\n",
    "function(atom,viewer) {\n",
    "    if(!atom.label) {\n",
    "        atom.label = viewer.addLabel(atom.interaction,\n",
    "            {position: atom, backgroundColor: 'black', fontColor:'white'});\n",
    "    }\n",
    "}\"\"\"\n",
    "unhover_func = \"\"\"\n",
    "function(atom,viewer) {\n",
    "    if(atom.label) {\n",
    "        viewer.removeLabel(atom.label);\n",
    "        delete atom.label;\n",
    "    }\n",
    "}\"\"\"\n",
    "\n",
    "v = py3Dmol.view(650, 600)\n",
    "v.removeAllModels()\n",
    "\n",
    "models = {}\n",
    "mid = -1\n",
    "for i, row in df.T.iterrows():\n",
    "    lresid, presid, interaction = i\n",
    "    lindex, pindex = row[0]\n",
    "    lres = lmol[lresid]\n",
    "    pres = pmol[presid]\n",
    "    # set model ids for reusing later\n",
    "    for resid, res, style in [\n",
    "        (lresid, lres, {\"colorscheme\": \"cyanCarbon\"}),\n",
    "        (presid, pres, {}),\n",
    "    ]:\n",
    "        if resid not in models.keys():\n",
    "            mid += 1\n",
    "            v.addModel(Chem.MolToMolBlock(res), \"sdf\")\n",
    "            model = v.getModel()\n",
    "            model.setStyle({}, {\"stick\": style})\n",
    "            # add residue label\n",
    "            model.setHoverable({}, True, resid_hover.format(resid), unhover_func)\n",
    "            models[resid] = mid\n",
    "    # get coordinates for both points of the interaction\n",
    "    if interaction in [\"PiStacking\", \"EdgeToFace\", \"FaceToFace\", \"PiCation\"]:\n",
    "        p1 = get_ring_centroid(lres, lindex)\n",
    "    else:\n",
    "        p1 = lres.GetConformer().GetAtomPosition(lindex)\n",
    "    if interaction in [\"PiStacking\", \"EdgeToFace\", \"FaceToFace\", \"CationPi\"]:\n",
    "        p2 = get_ring_centroid(pres, pindex)\n",
    "    else:\n",
    "        p2 = pres.GetConformer().GetAtomPosition(pindex)\n",
    "    # add interaction line\n",
    "    v.addCylinder(\n",
    "        {\n",
    "            \"start\": dict(x=p1.x, y=p1.y, z=p1.z),\n",
    "            \"end\": dict(x=p2.x, y=p2.y, z=p2.z),\n",
    "            \"color\": colors.get(interaction, \"grey\"),\n",
    "            \"radius\": 0.15,\n",
    "            \"dashed\": True,\n",
    "            \"fromCap\": 1,\n",
    "            \"toCap\": 1,\n",
    "        }\n",
    "    )\n",
    "    # add label when hovering the middle of the dashed line by adding a dummy atom\n",
    "    c = Geometry.Point3D(*plf.utils.get_centroid([p1, p2]))\n",
    "    modelID = models[lresid]\n",
    "    model = v.getModel(modelID)\n",
    "    model.addAtoms(\n",
    "        [{\"elem\": \"Z\", \"x\": c.x, \"y\": c.y, \"z\": c.z, \"interaction\": interaction}]\n",
    "    )\n",
    "    model.setStyle({\"interaction\": interaction}, {\"clicksphere\": {\"radius\": 0.5}})\n",
    "    model.setHoverable({\"interaction\": interaction}, True, hover_func, unhover_func)\n",
    "\n",
    "# show protein\n",
    "mol = Chem.RemoveAllHs(pmol)\n",
    "pdb = Chem.MolToPDBBlock(mol, flavor=0x20 | 0x10)\n",
    "v.addModel(pdb, \"pdb\")\n",
    "model = v.getModel()\n",
    "model.setStyle({}, {\"cartoon\": {\"style\": \"edged\"}})\n",
    "\n",
    "v.zoomTo({\"model\": list(models.values())})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c97a5db1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from prolif.plotting.network import LigNetwork\n",
    "\n",
    "net = LigNetwork.from_ifp(\n",
    "    df,\n",
    "    lmol,\n",
    "    # replace with `kind=\"frame\", frame=0` for the other depiction\n",
    "    kind=\"aggregate\",\n",
    "    threshold=0.3,\n",
    "    rotation=270,\n",
    ")\n",
    "net.display()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6d2c183",
   "metadata": {},
   "source": [
    "## Docking-based deep learning\n",
    "\n",
    "Deep learning can be used to improve molecular docking or boost molecular property prediction models. Examples are listed below.\n",
    "\n",
    "Neural networks for molecular docking:\n",
    "\n",
    "1. Lu, Wei, et al. \"Tankbind: Trigonometry-aware neural networks for drug-protein binding structure prediction.\" bioRxiv (2022): 2022-06.\n",
    "2. Stärk, Hannes, et al. \"Equibind: Geometric deep learning for drug binding structure prediction.\" International Conference on Machine Learning. PMLR, 2022.\n",
    "3. Corso, Gabriele, et al. \"Diffdock: Diffusion steps, twists, and turns for molecular docking.\" arXiv preprint arXiv:2210.01776 (2022).\n",
    "\n",
    "![](https://hannes-stark.com/assets/diffdock_example.gif)\n",
    "\n",
    "Binding affinity prediction using neural networks:\n",
    "\n",
    "1. McNutt, Andrew T., et al. \"GNINA 1.0: molecular docking with deep learning.\" Journal of cheminformatics 13.1 (2021): 1-20.\n",
    "2. Jones, Derek, et al. \"Improved protein–ligand binding affinity prediction with structure-based deep fusion inference.\" Journal of chemical information and modeling 61.4 (2021): 1583-1592."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
